### PR TITLE
Add testing for MNIST commands

### LINKED ISSUE #30: Support MNIST commands in threepio.js 💭
# Summary

With the newly introduced demo, we would like to support the commands provided in the demo. Lets add them to our test suite.

# Background

**Describe the ideal solution**:

Complete support for the below listed commands

# Details

We want to support the following commands

```
var_2 = W1.t()
var_4 = X.__matmul__(var_2)
var_6 = var_4.__add__(b1)
var_7 = torch.nn.functional.relu(var_6)
var_9 = W2.t()
var_10 = var_7.__matmul__(var_9)
var_12 = var_10.__add__(b2)
var_13 = torch.nn.functional.softmax(var_12, {'dim': 1})
var_14 = torch.log(var_13)
var_16 = y.__mul__(var_14)
var_17 = var_16.mean()
loss = var_17.__neg__()
var_19 = var_13.__sub__(y)
var_21 = batch_size.__mul__(10)
var_22 = var_19.__truediv__(var_21)
var_23 = var_22.__matmul__(W2)
var_24 = var_6.__gt__(0)
var_25 = var_24.float()
var_26 = var_23.__mul__(var_25)
var_27 = var_26.t()
var_28 = var_27.__matmul__(X)
var_29 = var_26.sum(0)
var_30 = var_22.t()
var_31 = var_30.__matmul__(var_7)
var_32 = var_22.sum(0)
var_34 = var_28.__mul__(lr)
upd_W1 = W1.__sub__(var_34)
var_36 = var_29.__mul__(lr)
upd_b1 = b1.__sub__(var_36)
var_38 = var_31.__mul__(lr)
upd_W2 = W2.__sub__(var_38)
var_40 = var_32.__mul__(lr)
upd_b2 = b2.__sub__(var_40)
var_42 = torch.argmax(var_13, {'dim': 1})
var_43 = torch.argmax(y, {'dim': 1})
var_44 = var_42.eq(var_43)
var_45 = var_44.float()
var_46 = var_45.sum()
acc = var_46.__truediv__(batch_size)
```

### PR DESCRIPTION
**Pull request by Nolski**
This PR should support all the commands needed to train a model on MNIST and resolves #30

### CODE REVIEW

File: js/src/threepio.js [hunk 1]
```diff
@@ -10,39 +11,34 @@ export default class Threepio {
...
-      // TODO: match args based on fromLibrary name
-      // TODO: Ensure order is preserved
-      // TODO: Ensure there are not any missing required arguments
-      console.log(i, arg);
-      continue;
+    if (result in this.mappedCommands[toLang]) {
+      return result;
+    }
+
+    if (result in WORD_TRANSLATIONS) {
```
Comment (vvmnnnkv) [inline js/src/threepio.js]: It's strange that when checking against mappedCommands, we look at target framework (`toLang`),
but WORD_TRANSLATIONS seem to be generic, and also having opposite translations like
```
  eq: 'equal',
  equal: 'eq',
```
It seems only one of these translations is true, depending on toLang? :)
Also, it seems strange that we need separate constant for this, why it's not in mappedCommands?

File: js/test/fixtures/pytorch.js [hunk 2]
```diff
@@ -4,82 +4,109 @@ import Command from '../../src/command';
...
+    tf.tensor([
+      [1, 0],
+      [0, 1]
+    ]),
+    tf.tensor([
+      [0, 1],
+      [1, 0]
+    ])
+  ],
   [['out', 'None']]
```
Comment (vvmnnnkv) [inline js/test/fixtures/pytorch.js]: What does out None mean?

File: js/test/fixtures/pytorch.js [hunk 3]
```diff
@@ -4,82 +4,109 @@ import Command from '../../src/command';
...
-  [tf.tensor([1]), 1],
+  [tf.tensor([1, 2, 3])],
   [['out', 'None']]
 );
 
 export const argmax = new Command(
   'torch.argmax(input, dim)',
   'argmax',
-  [tf.tensor([1]), 1],
+  [tf.tensor([0, 10, 2])],
```
Comment (vvmnnnkv) [inline js/test/fixtures/pytorch.js]: In mnist sample, argmax has kwarg dim=1
It would be interesting to test this works.

File: js/test/fixtures/pytorch.js [hunk 4]
```diff
@@ -4,82 +4,109 @@ import Command from '../../src/command';
...
+    ])
+  ],
   [['out', 'None']]
 );
 
 export const softmax = new Command(
   'torch.softmax(input, dim=None, other)',
   'softmax',
-  [tf.tensor([1]), 1],
+  [tf.tensor([2, 1, 0.1])],
```
Comment (vvmnnnkv) [inline js/test/fixtures/pytorch.js]: In mnist sample, softmax has kwarg dim=1
It would be interesting to test this works.

File: js/src/threepio.js [hunk 5]
```diff
@@ -10,39 +11,34 @@ export default class Threepio {
...
+      this._normalizeFunctionName(cmd.functionName, toLang)
     ];
 
     let translatedCmd = this[cmdInfo.attrs.shift()];
     while (cmdInfo.attrs.length > 0) {
       translatedCmd = translatedCmd[cmdInfo.attrs.shift()];
     }
 
-    translatedCmd(...cmd.args);
+    cmd.execFn = translatedCmd;
```
Comment (vvmnnnkv) [inline js/src/threepio.js]: Hmm, has side-effect on cmd outside of this method, is that intended?
(I.e. returning cmd is not necessary in this case)

File: js/src/command.js [hunk 6]
```diff
@@ -11,9 +10,14 @@ export default class Command {
...
   }
 
-  execute() {
-    tf[this.functionPath](...this.args);
+  executeRoutine() {
+    if (typeof this.execFn === 'undefined') {
+      throw new TypeError(NOT_TRANSLATED);
+    }
+
+    return this.execFn(...this.args);
```
Comment (vvmnnnkv) [inline js/src/command.js:20]: Hmm, what about kwargs?
E.g. `torch.nn.functional.softmax(tensor, dim=1)` (where dim is kwarg) becomes `tfjs.softmax(tensor, 1)` where dim is arg.

File: js/src/threepio.js [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import * as tf from '@tensorflow/tfjs';
```
Comment (vvmnnnkv) [inline js/src/threepio.js]: I wonder if it's possible to remove that import.
We could pass `tf` into `Threepio()` constructor.
Also user code can just execute something like `tf[cmd.execFnName](...cmd.args)`.

File: js/test/threepio.test.js [hunk 8]
```diff
@@ -1,23 +1,124 @@
...
   test('returns itself', () => {
     expect(threepio).toBeInstanceOf(Threepio);
   });
 
-  test('translate basic function', () => {
-    threepio.translate('tfjs', add);
+  test('translates add', () => {
+    const translation = threepio.translate('tfjs', add);
+    const result = translation.executeRoutine();
+    const answer = tf.tensor([
```
Comment (vvmnnnkv) [inline js/test/threepio.test.js]: Suggestion: it might be easier to keep `answer` (like `add_answer`) together with the command in fixtures.

File: js/src/threepio.js [hunk 9]
```diff
@@ -10,39 +11,34 @@ export default class Threepio {
...
-    argMap.args.sort((a, b) =>
-      a.index === b.index ? 0 : a.index < b.index ? -1 : 1
-    );
-    for (const [i, arg] of args.entries()) {
-      // TODO: match args based on fromLibrary name
-      // TODO: Ensure order is preserved
-      // TODO: Ensure there are not any missing required arguments
-      console.log(i, arg);
-      continue;
+    if (result in this.mappedCommands[toLang]) {
```
Comment (vvmnnnkv) [inline js/src/threepio.js]: Hmm, does it expect function names to be same in both frameworks?
E.g. func name in torch must directly match to func name in tfjs?
I imagined mapping will take in consideration source and target frameworks, like
```
torchCmdInfo = mappedCommands['torch'][name] // source framework
tfjsName = torchCmdInfo.how_that_will_be_in['tfjs']
tfjsCmdInfo = mappedCommands['tfjs'][tfjsName] // target framework
// ... args mapping ...
```

File: js/test/threepio.test.js (hunk 8 above)
Comment (vvmnnnkv) [reply js/test/threepio.test.js]: Suggestion #2: you can even do array of ops/answers in fixtures, and one test to loop over this array :)

File: js/src/threepio.js (hunk 1 above)
Comment (Nolski) [reply js/src/threepio.js]: I left it out of mappedCommands as this was meant as really word level translations. These translations will be used for argument mapping and others as sometimes certain libraries lexicon is structurally the same but have different words.

For instance, pytorch prefers `eq` for describing equality where-as tensorflow prefers `equal`.
Or for many tensor operations, pytorch prefers `input`, and `other` as the two tensors fed in. Tensorflow prefers `a` and `b`

File: js/test/fixtures/pytorch.js (hunk 3 above)
Comment (Nolski) [reply js/test/fixtures/pytorch.js]: `dim` doesn't yet automatically map to the tensorflow word which is `axis`

File: js/src/threepio.js (hunk 7 above)
Comment (Nolski) [reply js/src/threepio.js]: We could! Would this make things easier?

File: docs-crawler/docs/aggregate_crawler_output.py [hunk 10]
```diff
@@ -40,13 +46,17 @@ def load_base_defs(self):
         self.populate_command('torch')
 
     def hydrate_args(self, base_args, base_kwargs):
+        base_args = list(filter(lambda a: a not in ['', '?'], base_args))
```
Comment (Nolski) [inline docs-crawler/docs/aggregate_crawler_output.py:56]: We had lots of arguments with empty named and only question marks due to strange formatting of a few commands on the page. I am just stripping these out for now. In the future, we will want to just remove all support for these commands.

File: docs-crawler/docs/aggregate_crawler_output.py [hunk 11]
```diff
@@ -1,6 +1,18 @@
 import json
 import re
 
+TRANSLATIONS = {
```
Comment (Nolski) [inline docs-crawler/docs/aggregate_crawler_output.py:4]: So these are the translations that I moved into our crawler which will do word level translations for arguments & function names.

File: docs-crawler/docs/aggregate_crawler_output.py [hunk 12]
```diff
@@ -65,32 +81,37 @@ def match_arg_names(self, base, match, to_lang):
...
+                    base_arg['name'], None)
         return base
 
     def load_translations(self, from_lang):
         langs = ['torch', 'tfjs', 'tf']
         langs.pop(langs.index(from_lang))
 
         for d in self.base_defs:
+            # First perform any word level translations
+            from_d = TRANSLATIONS[from_lang].get(d, False) or d
```
Comment (Nolski) [inline docs-crawler/docs/aggregate_crawler_output.py:95]: Here is where the word translation happens and gets added into the json

File: js/src/command.js (hunk 6 above)
Comment (Nolski) [reply js/src/command.js:20]: Translation will modify `args` to place the kwargs in the right spot

File: js/src/threepio.js [hunk 13]
```diff
@@ -1,48 +1,79 @@
...
-    argMap.args.sort((a, b) =>
-      a.index === b.index ? 0 : a.index < b.index ? -1 : 1
-    );
-    for (const [i, arg] of args.entries()) {
-      // TODO: match args based on fromLibrary name
-      // TODO: Ensure order is preserved
-      // TODO: Ensure there are not any missing required arguments
-      console.log(i, arg);
-      continue;
+  orderArgs(cmd, fromInfo, toInfo) {
```
Comment (Nolski) [inline js/src/threepio.js]: This is the logic for parsing the location of the all of the args and kwargs, and re-ordering them to match the formatting of the command we are translating to.

File: js/test/fixtures/pytorch.js [hunk 14]
```diff
@@ -1,85 +1,175 @@
...
 import Command from '../../src/command';
 
-export const add = new Command(
-  'torch.add(input,other,out=None',
-  'add',
-  [tf.tensor([1]), tf.tensor([1])],
-  [['out', 'None']]
-);
+export const add = {
+  inputs: [
```
Comment (Nolski) [inline js/test/fixtures/pytorch.js:10]: East text fixture has `inputs` and `answers` (of the same length) which will be compared to each other in our test suite.

File: js/test/threepio.test.js [hunk 15]
```diff
@@ -1,23 +1,96 @@
...
+    processTests(argmax);
+  });
+
+  // TODO: Write tests for kwargs use of dim
+
+  test('translates transpose', () => {
+    processTests(t);
+  });
+
+  test('translates softmax', () => {
```
Comment (Nolski) [inline js/test/threepio.test.js:90]: softmax requires a special test as we need to use `toBeCloseTo` due to rounding errors

File: js/src/threepio.js (hunk 9 above)
Comment (Nolski) [reply js/src/threepio.js]: We now store the function name mapping in our static json.

File: js/src/threepio.js (hunk 7 above)
Comment (vvmnnnkv) [reply js/src/threepio.js]: We require only tf-core in syft.js, it's smaller part of tfjs.
This import requires syft.js to use full tfjs :)

**Review by vvmnnnkv — APPROVED** [review]

File: js/src/threepio.js (hunk 7 above)
Comment (cereallarceny) [reply js/src/threepio.js]: @Nolski We need to do this. Good idea @vvmnnnkv

Comment (cereallarceny) [reply js/src/threepio.js]: Rather than even do that, you could just assume that the `tf` object is a global that will exist. So long as it's being imported for syft.js successfully, threepio can assume it's there. Should be able to do: `/* global tf */` at the top of the file.

**Review by cereallarceny — APPROVED** [review]

**Comment by cereallarceny** [comment]
Great work @Nolski! If you could just get that issue with importing `tf` correctly, that's my only issue before merging!
